### PR TITLE
GEOT-6207: Better disposal of inner images through ImageUtilities dis…

### PR DESCRIPTION
**Pull request by dromagnoli**
…posal methods

### CODE REVIEW

**Comment by dromagnoli** [comment]
coverage disposal has been removed from tests since the underlying image get already disposed.
rendering test has been updated to avoid re-using an already disposed image

File: modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java [hunk 1]
```diff
@@ -1276,11 +1291,12 @@ public static void disposeSinglePlanarImage(PlanarImage planarImage) {
...
 
         try {
-            if (planarImage instanceof RenderedImageAdapter) {
-                cleanField(planarImage, "theImage");
-            }
+            // Check extended class first
             if (planarImage instanceof WritableRenderedImageAdapter) {
                 cleanField(planarImage, "theWritableImage");
+                cleanField(planarImage, "theImage", true);
+            } else if (planarImage instanceof RenderedImageAdapter) {
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: Before the two ifs were not alternative, now they are. I guess this is correct as the two fields in question are pointing to the same image right?

File: modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java [hunk 2]
```diff
@@ -1303,10 +1319,30 @@ public static void disposeSinglePlanarImage(PlanarImage planarImage) {
...
+     * @param theObject
+     * @param fieldName
+     * @throws NoSuchFieldException
+     * @throws IllegalAccessException
+     */
+    private static void cleanField(Object theObject, String fieldName, boolean superClass)
+            throws NoSuchFieldException, IllegalAccessException {
+        // getDeclaredField only provides access to current class,
+        // not superclass.
+        Class<? extends Object> theClass = theObject.getClass();
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: Why not use the getField method instead? It goes up the inheritance chain by itself, no need for flags

File: modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java [hunk 3]
```diff
@@ -223,7 +223,7 @@ public void testDrawIntepolation() throws Exception {
         layer.getUserData()
                 .put(
                         StreamingRenderer.BYLAYER_INTERPOLATION,
-                        Interpolation.getInstance(Interpolation.INTERP_BICUBIC));
+                        Interpolation.getInstance(Interpolation.INTERP_BICUBIC_2));
```
Comment (aaime) [inline modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java]: What happened here?

File: modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java [hunk 4]
```diff
@@ -237,19 +237,9 @@ public void testDrawIntepolation() throws Exception {
         // test right interpolation hint is set on Graphics2D
         assertEquals(
                 graphics.getRenderingHint(JAI.KEY_INTERPOLATION),
-                Interpolation.getInstance(Interpolation.INTERP_BICUBIC));
```
Comment (aaime) [inline modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java]: And here?

File: modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java (hunk 2 above)
Comment (dromagnoli) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: getField only get public fields, not private ones :(

File: modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java (hunk 1 above)
Comment (dromagnoli) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: right.

File: modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java (hunk 4 above)
Comment (dromagnoli) [reply modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java]: Just using a "not commonly used" interpolation to make sure hints is actually using it once set (so it's not using a default)

File: modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java [hunk 5]
```diff
@@ -1276,11 +1291,12 @@ public static void disposeSinglePlanarImage(PlanarImage planarImage) {
...
         try {
-            if (planarImage instanceof RenderedImageAdapter) {
-                cleanField(planarImage, "theImage");
-            }
+            // Check extended class first
             if (planarImage instanceof WritableRenderedImageAdapter) {
                 cleanField(planarImage, "theWritableImage");
+                cleanField(planarImage, "theImage", true);
+            } else if (planarImage instanceof RenderedImageAdapter) {
+                cleanField(planarImage, "theImage");
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: @lislei do you have any chance to setup a dataset (data + style eventually and configuration steps) and requests to replicate that?

[the remark this answers is not in the document]
Comment (lislei) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: What we have is a webstart app based on geotools that produces this nullpointer when painting a wms layer. I could look into giving you access to it if that would help?

[the remark this answers is not in the document]
Comment (lislei) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: My colegue pointed out to me [a self contained example in this post](https://sourceforge.net/p/geotools/mailman/message/36602879/) - see attachement "WmsTest.java". 
This should produce the nullpointer.

[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java]: Your colleague deserves to pay and use commercial software only. If you are interested in getting any help instead, please open a ticket on https://osgeo-org.atlassian.net/projects/GEOT attaching the example and wording the description like a decent, polite person would do.

**Pull request merged**